File: app/schemas/supply_chain.py
"""产业供应链图谱 API 请求/响应 Schema。"""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.models.graph_entity import EntityType
from app.models.graph_fact import RelationType
from app.models.graph_source import DocumentType


# ──────────────────────────────────────────────
# 摄取（Ingest）
# ──────────────────────────────────────────────


class IngestDocumentRequest(BaseModel):
    """手动触发文档摄取请求。"""

    url: str = Field(description="文档 URL，SEC EDGAR 直链或电话会议文字记录")
    document_type: DocumentType
    ticker: Optional[str] = Field(default=None, description="相关股票代码（如 NVDA）")
    company_name: Optional[str] = Field(default=None)
    filing_date: Optional[datetime] = Field(default=None)
    period_of_report: Optional[datetime] = Field(default=None)
    section: Optional[str] = Field(default=None, description="文档章节（如 Risk Factors）")
    title: Optional[str] = Field(default=None)


class IngestDocumentResponse(BaseModel):
    """文档摄取响应。"""

    doc_id: uuid.UUID
    status: str
    message: str


class IngestTextRequest(BaseModel):
    """直接摄取原始文本请求。"""

    text: str = Field(description="原始文本内容（电话会议记录、IR 材料等）", min_length=100)
    document_type: DocumentType
    ticker: Optional[str] = Field(default=None, description="相关股票代码（如 NVDA）")
    company_name: Optional[str] = Field(default=None)
    period_of_report: Optional[datetime] = Field(default=None, description="报告期 YYYY-MM-DD")
    section: Optional[str] = Field(default=None)
    title: Optional[str] = Field(default=None)


# ──────────────────────────────────────────────
# 实体（Entity）
# ──────────────────────────────────────────────


class EntityOut(BaseModel):
    """实体节点输出 Schema。"""

    id: uuid.UUID
    entity_type: EntityType
    name: str
    aliases: list[str]
    description: Optional[str]
    ticker: Optional[str]
    created_at: datetime

    model_config = {"from_attributes": True}


class EntityCreate(BaseModel):
    """手动创建实体请求。"""

    entity_type: EntityType
    name: str
    aliases: list[str] = Field(default_factory=list)
    description: Optional[str] = None
    ticker: Optional[str] = None


# ──────────────────────────────────────────────
# 事实（Fact）
# ──────────────────────────────────────────────


class FactOut(BaseModel):
    """事实关系输出 Schema。"""

    id: uuid.UUID
    source_entity_id: uuid.UUID
    target_entity_id: uuid.UUID
    source_entity_name: Optional[str] = None  # 通过 join 填充
    target_entity_name: Optional[str] = None
    source_entity_type: Optional[EntityType] = None
    target_entity_type: Optional[EntityType] = None
    relation_type: RelationType
    evidence_text: str
    confidence: float
    event_time: Optional[datetime]
    ingestion_time: datetime
    document_url: Optional[str]
    document_section: Optional[str]
    chunk_id: Optional[str]

    model_config = {"from_attributes": True}


class FactCreate(BaseModel):
    """手动创建事实请求。"""

    source_entity_id: uuid.UUID
    target_entity_id: uuid.UUID
    relation_type: RelationType
    evidence_text: str
    confidence: float = 0.8
    event_time: Optional[datetime] = None
    document_url: Optional[str] = None
    document_section: Optional[str] = None


# ──────────────────────────────────────────────
# 图谱查询
# ──────────────────────────────────────────────


class GraphNode(BaseModel):
    """前端图谱节点数据。"""

    id: str
    name: str
    entity_type: EntityType
    ticker: Optional[str]
    description: Optional[str]
    fact_count: int = 0


class GraphEdge(BaseModel):
    """前端图谱边数据。"""

    id: str
    source: str
    target: str
    relation_type: RelationType
    evidence_text: str
    confidence: float
    event_time: Optional[datetime]
    document_url: Optional[str]


class GraphData(BaseModel):
    """图谱可视化数据（节点 + 边）。"""

    nodes: list[GraphNode]
    edges: list[GraphEdge]
    total_entities: int
    total_facts: int


class GraphQueryParams(BaseModel):
    """图谱查询参数。"""

    entity_types: Optional[list[EntityType]] = None
    relation_types: Optional[list[RelationType]] = None
    ticker: Optional[str] = None
    min_confidence: float = 0.0
    limit: int = Field(default=200, ge=1, le=1000)


class BottleneckReport(BaseModel):
    """产业瓶颈分析报告。"""

    resource_name: str
    resource_type: EntityType
    constrained_count: int  # 受约束的实体数量
    constrained_entities: list[EntityOut]
    evidence_samples: list[str]


class DemandChain(BaseModel):
    """需求传导链路。"""

    concept: EntityOut
    enabled_products: list[EntityOut]
    supplier_companies: list[EntityOut]
    constrained_resources: list[EntityOut]

File: frontend/lib/api/supply_chain.ts
import apiClient from './client'

export type EntityType = 'Company' | 'Product' | 'Technology' | 'Concept' | 'Resource'
export type RelationType = 'HAS_PRODUCT' | 'SUPPLIED_BY' | 'ENABLED_BY' | 'CONSTRAINED_BY'
export type DocumentType = '10-K' | '10-Q' | '8-K' | 'earnings_call' | 'investor_relations'
export type DocumentStatus = 'pending' | 'processing' | 'done' | 'failed'

export interface Entity {
  id: string
  entity_type: EntityType
  name: string
  aliases: string[]
  description: string | null
  ticker: string | null
  created_at: string
}

export interface Fact {
  id: string
  source_entity_id: string
  target_entity_id: string
  source_entity_name: string | null
  target_entity_name: string | null
  source_entity_type: EntityType | null
  target_entity_type: EntityType | null
  relation_type: RelationType
  evidence_text: string
  confidence: number
  event_time: string | null
  ingestion_time: string
  document_url: string | null
  document_section: string | null
  chunk_id: string | null
}

export interface GraphNode {
  id: string
  name: string
  entity_type: EntityType
  ticker: string | null
  description: string | null
  fact_count: number
}

export interface GraphEdge {
  id: string
  source: string
  target: string
  relation_type: RelationType
  evidence_text: string
  confidence: number
  event_time: string | null
  document_url: string | null
}

export interface GraphData {
  nodes: GraphNode[]
  edges: GraphEdge[]
  total_entities: number
  total_facts: number
}

export interface BottleneckReport {
  resource_name: string
  resource_type: EntityType
  constrained_count: number
  constrained_entities: Entity[]
  evidence_samples: string[]
  description: string
}

export interface DemandChain {
  concept: Entity
  enabled_products: Entity[]
  supplier_companies: Entity[]
  constrained_resources: Entity[]
}

export interface GraphStats {
  entities: Record<EntityType, number>
  facts: Record<RelationType, number>
  total_entities: number
  total_facts: number
  documents: { total: number; done: number }
}

export interface SourceDoc {
  id: string
  url: string
  document_type: string
  ticker: string | null
  company_name: string | null
  status: DocumentStatus
  chunk_count: number
  fact_count: number
  created_at: string
  ingested_at: string | null
}

export interface IngestRequest {
  url: string
  document_type: DocumentType
  ticker?: string
  company_name?: string
  filing_date?: string
  period_of_report?: string
  section?: string
  title?: string
}

export interface IngestTextRequest {
  text: string
  document_type: DocumentType
  ticker?: string
  company_name?: string
  period_of_report?: string
  section?: string
  title?: string
}

const BASE = '/api/v1/supply-chain'

export const supplyChainApi = {
  getGraph: async (params?: {
    entity_types?: string
    relation_types?: string
    ticker?: string
    min_confidence?: number
    since?: string
    limit?: number
  }): Promise<GraphData> => {
    const { data } = await apiClient.get(`${BASE}/graph`, { params })
    return data
  },

  getStats: async (): Promise<GraphStats> => {
    const { data } = await apiClient.get(`${BASE}/stats`)
    return data
  },

  getEntities: async (params?: {
    entity_type?: EntityType
    ticker?: string
    search?: string
    limit?: number
  }): Promise<Entity[]> => {
    const { data } = await apiClient.get(`${BASE}/entities`, { params })
    return data
  },

  createEntity: async (body: {
    entity_type: EntityType
    name: string
    aliases?: string[]
    description?: string
    ticker?: string
  }): Promise<Entity> => {
    const { data } = await apiClient.post(`${BASE}/entities`, body)
    return data
  },

  getEntityFacts: async (id: string, direction?: 'inbound' | 'outbound' | 'both'): Promise<Fact[]> => {
    const { data } = await apiClient.get(`${BASE}/entities/${id}/facts`, {
      params: { direction },
    })
    return data
  },

  getFacts: async (params?: {
    relation_type?: RelationType
    min_confidence?: number
    doc_id?: string
    limit?: number
  }): Promise<Fact[]> => {
    const { data } = await apiClient.get(`${BASE}/facts`, { params })
    return data
  },

  createFact: async (body: {
    source_entity_id: string
    target_entity_id: string
    relation_type: RelationType
    evidence_text: string
    confidence?: number
    event_time?: string
    document_url?: string
    document_section?: string
  }): Promise<Fact> => {
    const { data } = await apiClient.post(`${BASE}/facts`, body)
    return data
  },

  getBottlenecks: async (): Promise<BottleneckReport[]> => {
    const { data } = await apiClient.get(`${BASE}/analysis/bottlenecks`)
    return data
  },

  getDemandChain: async (concept: string): Promise<DemandChain> => {
    const { data } = await apiClient.get(`${BASE}/analysis/demand-chain/${encodeURIComponent(concept)}`)
    return data
  },

  getDocuments: async (params?: {
    ticker?: string
    document_type?: string
    status?: DocumentStatus
    limit?: number
  }): Promise<SourceDoc[]> => {
    const { data } = await apiClient.get(`${BASE}/documents`, { params })
    return data
  },

  ingestDocument: async (body: IngestRequest): Promise<{ doc_id: string; status: string; message: string }> => {
    const { data } = await apiClient.post(`${BASE}/ingest`, body)
    return data
  },

  ingestText: async (body: IngestTextRequest): Promise<{ doc_id: string; status: string; message: string }> => {
    const { data } = await apiClient.post(`${BASE}/ingest/text`, body)
    return data
  },

  ingestSec: async (params: {
    ticker: string
    form_type?: string
    section?: string
  }): Promise<{ status: string; message: string }> => {
    const { data } = await apiClient.post(`${BASE}/ingest/sec`, null, { params })
    return data
  },

  ingestEarningsCall: async (params: {
    ticker: string
    year: number
    quarter: number
  }): Promise<{ status: string; message: string }> => {
    const { data } = await apiClient.post(`${BASE}/ingest/earnings-call`, null, { params })
    return data
  },

  searchSecFilings: async (params: {
    ticker: string
    form_types?: string
    start_date?: string
    end_date?: string
    max_results?: number
  }): Promise<{ ticker: string; results: Array<{
    form: string
    filing_date: string
    period_of_report: string
    entity_name: string
    ticker: string
    index_url: string | null
  }>; count: number }> => {
    const { data } = await apiClient.get(`${BASE}/ingest/sec/search`, { params })
    return data
  },

  listEarningsCallAvailable: async (ticker: string): Promise<{
    ticker: string
    available: Array<{ year: number; quarter: number; date: string }>
    count: number
  }> => {
    const { data } = await apiClient.get(`${BASE}/ingest/earnings-call/list`, { params: { ticker } })
    return data
  },
}
